File: bot/src/domain/shared/market.py
"""Market domain value objects"""
from dataclasses import dataclass
from typing import Optional


@dataclass(frozen=True)
class TradeGood:
    """
    Trade good value object

    Represents a single commodity available at a market.
    Prices follow the market's perspective:
    - purchase_price: What the market PAYS when buying from ships (market bids)
    - sell_price: What the market CHARGES when selling to ships (market asks)
    """
    symbol: str                      # Commodity symbol (e.g., "IRON_ORE")
    supply: Optional[str]            # SCARCE, LIMITED, MODERATE, HIGH, ABUNDANT
    activity: Optional[str]          # WEAK, GROWING, STRONG, RESTRICTED
    purchase_price: int              # What ship RECEIVES when selling to market
    sell_price: int                  # What ship PAYS when buying from market
    trade_volume: int                # Trading volume


@dataclass(frozen=True)
class Market:
    """
    Market snapshot value object

    Immutable snapshot of market data at a specific time.
    """
    waypoint_symbol: str                   # Waypoint identifier
    trade_goods: tuple[TradeGood, ...]     # Immutable tuple of trade goods
    last_updated: str                      # ISO timestamp


@dataclass(frozen=True)
class TourResult:
    """Result from executing one market tour"""
    markets_visited: int
    goods_updated: int
    duration_seconds: float


@dataclass(frozen=True)
class PollResult:
    """Result from executing one stationary poll"""
    goods_updated: int
    waypoint: str

File: gobot/internal/domain/trading/market.go
package trading

import "fmt"

// TradeGood is a value object representing a tradeable good at a market
type TradeGood struct {
	Symbol        string
	Supply        string
	SellPrice     int
	PurchasePrice int
	TradeVolume   int
}

// Market is a domain entity representing a marketplace
type Market struct {
	waypointSymbol string
	tradeGoods     []TradeGood
}

// NewMarket creates a new market
func NewMarket(waypointSymbol string, tradeGoods []TradeGood) (*Market, error) {
	if waypointSymbol == "" {
		return nil, fmt.Errorf("waypoint symbol cannot be empty")
	}

	return &Market{
		waypointSymbol: waypointSymbol,
		tradeGoods:     tradeGoods,
	}, nil
}

// WaypointSymbol returns the waypoint symbol
func (m *Market) WaypointSymbol() string {
	return m.waypointSymbol
}

// TradeGoods returns all trade goods
func (m *Market) TradeGoods() []TradeGood {
	return m.tradeGoods
}

// GetTradeGood finds a specific trade good by symbol
func (m *Market) GetTradeGood(symbol string) (*TradeGood, bool) {
	for i := range m.tradeGoods {
		if m.tradeGoods[i].Symbol == symbol {
			return &m.tradeGoods[i], true
		}
	}
	return nil, false
}

// GetTransactionLimit returns the trade volume for a good (or 999999 if not found)
func (m *Market) GetTransactionLimit(symbol string) int {
	good, found := m.GetTradeGood(symbol)
	if !found {
		return 999999
	}
	return good.TradeVolume
}

// HasGood checks if market sells a specific good
func (m *Market) HasGood(symbol string) bool {
	_, found := m.GetTradeGood(symbol)
	return found
}
